File: api/app/api/endpoints/injection.js
const express = require('express');
const router = express.Router();
const puppeteer = require('puppeteer');

/**
 * Memory Injection API
 * Handles injecting processed conversations into ChatGPT memory
 */

// Start memory injection process
router.post('/', async (req, res) => {
  const authHeader = req.headers.authorization;
  
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ 
      success: false, 
      message: 'No authentication token provided' 
    });
  }
  
  const token = authHeader.split(' ')[1];
  const { conversation_ids, config } = req.body;
  
  if (!conversation_ids || !Array.isArray(conversation_ids) || conversation_ids.length === 0) {
    return res.status(400).json({
      success: false,
      message: 'At least one conversation ID is required'
    });
  }
  
  // Generate a task ID for tracking
  const taskId = `inject_${Date.now()}_${Math.random().toString(36).substring(2, 10)}`;
  
  // Start the injection process in the background
  injectMemoryBackground(token, conversation_ids, config, taskId);
  
  // Return the task ID for status checking
  return res.status(202).json({
    success: true,
    message: 'Memory injection started',
    taskId
  });
});

// Check injection status
router.get('/:taskId', async (req, res) => {
  const { taskId } = req.params;
  
  // In a real implementation, this would check a database or cache
  // For this demo, we'll simulate status updates
  const status = getInjectionStatus(taskId);
  
  return res.status(200).json({
    success: true,
    taskId,
    status
  });
});

// Background injection process
async function injectMemoryBackground(token, conversationIds, config, taskId) {
  try {
    // Launch browser
    const browser = await puppeteer.launch({
      headless: true,
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    
    const page = await browser.newPage();
    
    // Set session cookie
    await page.setCookie({
      name: '__Secure-next-auth.session-token',
      value: token,
      domain: 'chat.openai.com',
      path: '/',
      httpOnly: true,
      secure: true
    });
    
    // Navigate to ChatGPT
    await page.goto('https://chat.openai.com/', { waitUntil: 'networkidle2' });
    
    // Wait for page to load
    await page.waitForSelector('[data-testid="text-input-button"]', { timeout: 10000 }).catch(() => {});
    
    // For each conversation, create a memory injection message
    for (const conversationId of conversationIds) {
      // Update status
      updateInjectionStatus(taskId, {
        status: 'processing',
        current: conversationIds.indexOf(conversationId) + 1,
        total: conversationIds.length,
        message: `Injecting conversation ${conversationId}`
      });
      
      // Get conversation details
      const conversation = await fetchConversationDetails(page, conversationId);
      
      if (!conversation) {
        console.error(`Failed to fetch conversation ${conversationId}`);
        continue;
      }
      
      // Format the conversation for memory injection
      const memoryText = formatConversationForMemory(conversation, config);
      
      // Type the memory injection message
      await page.type('[data-testid="text-input"]', memoryText);
      
      // Send the message
      await page.click('[data-testid="text-input-button"]');
      
      // Wait for response
      await page.waitForSelector('[data-testid="conversation-turn"]', { timeout: 30000 }).catch(() => {});
      
      // Wait a bit before the next injection
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
    
    // Update status to completed
    updateInjectionStatus(taskId, {
      status: 'completed',
      message: 'Memory injection completed successfully'
    });
    
    // Close browser
    await browser.close();
    
  } catch (error) {
    console.error('Memory injection error:', error);
    
    // Update status to failed
    updateInjectionStatus(taskId, {
      status: 'failed',
      message: `Memory injection failed: ${error.message || 'Unknown error'}`
    });
  }
}

// Fetch conversation details
async function fetchConversationDetails(page, conversationId) {
  try {
    return await page.evaluate(async (convId) => {
      const response = await fetch(`https://chat.openai.com/backend-api/conversation/${convId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json'
        }
      });
      
      if (!response.ok) {
        throw new Error('Failed to fetch conversation details');
      }
      
      return await response.json();
    }, conversationId);
  } catch (error) {
    console.error(`Error fetching conversation ${conversationId}:`, error);
    return null;
  }
}

// Format conversation for memory injection
function formatConversationForMemory(conversation, config) {
  let memoryText = `Please remember the following conversation from ${new Date(conversation.create_time * 1000).toLocaleString()}:\n\n`;
  
  // Add title
  memoryText += `Title: ${conversation.title}\n\n`;
  
  // Add messages
  if (conversation.mapping) {
    const rootNode = conversation.current_node;
    const messageIds = [];
    
    // Build message chain from the root node
    let currentId = rootNode;
    while (currentId && conversation.mapping[currentId]) {
      const node = conversation.mapping[currentId];
      if (node.message) {
        messageIds.unshift(currentId);
      }
      currentId = node.parent;
    }
    
    // Format messages
    for (const id of messageIds) {
      const messageData = conversation.mapping[id].message;
      if (messageData) {
        const role = messageData.author.role === 'assistant' ? 'ChatGPT' : 'User';
        const content = messageData.content.parts.join('\n');
        memoryText += `${role}: ${content}\n\n`;
      }
    }
  }
  
  // Add memory instruction
  memoryText += "Please confirm you've stored this conversation in your memory.";
  
  return memoryText;
}

// In-memory storage for injection status (would be a database in production)
const injectionStatus = {};

// Update injection status
function updateInjectionStatus(taskId, status) {
  injectionStatus[taskId] = {
    ...status,
    updatedAt: new Date().toISOString()
  };
}

// Get injection status
function getInjectionStatus(taskId) {
  return injectionStatus[taskId] || {
    status: 'unknown',
    message: 'Task not found'
  };
}

module.exports = router;

File: api/app/api/endpoints/injection.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks, Body
from typing import List, Dict, Any
import uuid
import asyncio

from app.api.endpoints.auth import get_current_user, User
from app.models.schemas import InjectionConfig, InjectionRequest, InjectionStatus, ProcessingStatus

router = APIRouter(tags=["injection"])

# Simulated injection tasks database - in production, use a real database
injection_tasks = {}


async def inject_memory_task(task_id: str, conversation_ids: List[str], config: InjectionConfig):
    """
    Background task to inject memory into ChatGPT
    
    Args:
        task_id: Task ID
        conversation_ids: List of conversation IDs to inject
        config: Injection configuration
    """
    # Update task status to processing
    injection_tasks[task_id]["status"] = ProcessingStatus.PROCESSING
    
    total_conversations = len(conversation_ids)
    successful = 0
    failed = 0
    
    # Simulate processing with progress updates
    for i, conv_id in enumerate(conversation_ids):
        # Simulate processing time
        await asyncio.sleep(1)
        
        # Update progress
        progress = (i + 1) / total_conversations
        injection_tasks[task_id]["progress"] = progress
        injection_tasks[task_id]["message"] = f"Injecting conversation {i+1}/{total_conversations}"
        
        # Simulate injection with retry logic
        success = False
        for attempt in range(config.retry_attempts):
            # Simulate API call
            await asyncio.sleep(0.5)
            
            # Simulate success (90% chance)
            if uuid.uuid4().int % 10 != 0:
                success = True
                break
            
            # Simulate retry delay
            await asyncio.sleep(config.retry_delay)
        
        if success:
            successful += 1
        else:
            failed += 1
        
        # Update task with current stats
        injection_tasks[task_id]["successful_injections"] = successful
        injection_tasks[task_id]["failed_injections"] = failed
    
    # Complete the task
    injection_tasks[task_id]["status"] = ProcessingStatus.COMPLETED
    injection_tasks[task_id]["progress"] = 1.0
    injection_tasks[task_id]["message"] = "Injection completed"


@router.post("/inject", response_model=InjectionStatus)
async def inject_memory(
    injection_request: InjectionRequest,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user)
):
    """
    Inject conversations into ChatGPT memory
    
    Args:
        injection_request: Injection request with conversation IDs and configuration
        background_tasks: FastAPI background tasks
        current_user: Current authenticated user
        
    Returns:
        InjectionStatus with task ID and initial status
    """
    # Create a new task
    task_id = str(uuid.uuid4())
    
    # Initialize task
    task = {
        "task_id": task_id,
        "status": ProcessingStatus.PENDING,
        "progress": 0.0,
        "message": "Task initialized",
        "successful_injections": 0,
        "failed_injections": 0
    }
    
    # Store task
    injection_tasks[task_id] = task
    
    # Start background processing
    background_tasks.add_task(
        inject_memory_task,
        task_id=task_id,
        conversation_ids=injection_request.conversation_ids,
        config=injection_request.config
    )
    
    return InjectionStatus(**task)


@router.get("/inject/{task_id}", response_model=InjectionStatus)
async def get_injection_status(
    task_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get status of an injection task
    
    Args:
        task_id: Task ID
        current_user: Current authenticated user
        
    Returns:
        InjectionStatus with current status
        
    Raises:
        HTTPException: If task not found
    """
    if task_id not in injection_tasks:
        raise HTTPException(status_code=404, detail="Task not found")
    
    return InjectionStatus(**injection_tasks[task_id])


@router.post("/inject/cancel/{task_id}")
async def cancel_injection_task(
    task_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Cancel an injection task
    
    Args:
        task_id: Task ID
        current_user: Current authenticated user
        
    Returns:
        Cancellation result
        
    Raises:
        HTTPException: If task not found or already completed
    """
    if task_id not in injection_tasks:
        raise HTTPException(status_code=404, detail="Task not found")
    
    task = injection_tasks[task_id]
    
    if task["status"] in [ProcessingStatus.COMPLETED, ProcessingStatus.FAILED]:
        raise HTTPException(status_code=400, detail="Task already completed or failed")
    
    # Update task status
    task["status"] = ProcessingStatus.FAILED
    task["message"] = "Task cancelled by user"
    
    return {"success": True, "message": "Task cancelled successfully"}
